From 44ecc520544df0fd0a188ef5907b7b99b965ce76 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 Mar 2021 14:16:52 -0500
Subject: [PATCH 055/196] Require PrecomputedTransactionData in
 MutableTransactionSignatureCreator. (#108)

This is so we can sign BIP341 signatures, which need the spent outputs, which is provided in PrecomputedTransactionData.
---
 src/bitcoin-tx.cpp                   | 10 ++++-
 src/psbt.cpp                         | 19 ++++++++-
 src/rpc/rawtransaction.cpp           | 23 +++++++----
 src/script/sign.cpp                  | 62 ++++++++++++++++++++--------
 src/script/sign.h                    |  8 ++--
 src/test/descriptor_tests.cpp        | 18 +++++---
 src/test/fuzz/script_sign.cpp        |  4 +-
 src/test/script_tests.cpp            |  5 ++-
 src/test/transaction_tests.cpp       | 27 +++++++-----
 src/test/txvalidationcache_tests.cpp | 18 ++++----
 10 files changed, 135 insertions(+), 59 deletions(-)

diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index c2f580c2e..87d09f6c1 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -683,6 +683,14 @@ static void MutateTxSign(CMutableTransaction &tx, const std::string &flagStr) {
 
     const FillableSigningProvider &keystore = tempKeystore;
 
+    // Collect spent outputs for BIP341 sighash
+    std::vector<CTxOut> spent_outputs;
+    spent_outputs.reserve(mergedTx.vin.size());
+    for (const CTxIn &txin : mergedTx.vin) {
+        spent_outputs.push_back(view.AccessCoin(txin.prevout).GetTxOut());
+    }
+    const PrecomputedTransactionData txdata(mergedTx, std::move(spent_outputs));
+
     // Sign what we can:
     for (size_t i = 0; i < mergedTx.vin.size(); i++) {
         CTxIn &txin = mergedTx.vin[i];
@@ -701,7 +709,7 @@ static void MutateTxSign(CMutableTransaction &tx, const std::string &flagStr) {
             (i < mergedTx.vout.size())) {
             ProduceSignature(keystore,
                              MutableTransactionSignatureCreator(
-                                 &mergedTx, i, amount, sigHashType),
+                                 &mergedTx, i, amount, sigHashType, txdata),
                              prevPubKey, sigdata);
         }
 
diff --git a/src/psbt.cpp b/src/psbt.cpp
index 0886b3458..e03497e2e 100644
--- a/src/psbt.cpp
+++ b/src/psbt.cpp
@@ -169,13 +169,21 @@ void UpdatePSBTOutput(const SigningProvider &provider,
     SignatureData sigdata;
     psbt_out.FillSignatureData(sigdata);
 
+    // Collect spent outputs for BIP341 sighash
+    std::vector<CTxOut> spent_outputs;
+    spent_outputs.reserve(psbt.inputs.size());
+    for (const PSBTInput &input : psbt.inputs) {
+        spent_outputs.push_back(input.utxo);
+    }
+    const PrecomputedTransactionData txdata(*psbt.tx, std::move(spent_outputs));
+
     // Construct a would-be spend of this output, to update sigdata with.
     // Note that ProduceSignature is used to fill in metadata (not actual
     // signatures), so provider does not need to provide any private keys (it
     // can be a HidingSigningProvider).
     MutableTransactionSignatureCreator creator(
         psbt.tx ? &psbt.tx.value() : nullptr, /* index */ 0, out.nValue,
-        SigHashType().withForkId());
+        SigHashType().withForkId(), txdata);
     ProduceSignature(provider, creator, out.scriptPubKey, sigdata);
 
     // Put redeem_script and key paths, into PSBTOutput.
@@ -211,8 +219,15 @@ bool SignPSBTInput(const SigningProvider &provider,
         sig_complete = ProduceSignature(provider, DUMMY_SIGNATURE_CREATOR,
                                         utxo.scriptPubKey, sigdata);
     } else {
+        // Collect spent outputs for BIP341 sighash
+        std::vector<CTxOut> spent_outputs;
+        spent_outputs.reserve(psbt.inputs.size());
+        for (const PSBTInput &input : psbt.inputs) {
+            spent_outputs.push_back(input.utxo);
+        }
+        const PrecomputedTransactionData txdata(tx, std::move(spent_outputs));
         MutableTransactionSignatureCreator creator(&tx, index, utxo.nValue,
-                                                   sighash);
+                                                   sighash, txdata);
         sig_complete =
             ProduceSignature(provider, creator, utxo.scriptPubKey, sigdata);
     }
diff --git a/src/rpc/rawtransaction.cpp b/src/rpc/rawtransaction.cpp
index 2b46a17a7..c6dd32440 100644
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -785,17 +785,24 @@ static UniValue combinerawtransaction(const Config &config,
     // Use CTransaction for the constant parts of the
     // transaction to avoid rehashing.
     const CTransaction txConst(mergedTx);
-    // Sign what we can:
-    for (size_t i = 0; i < mergedTx.vin.size(); i++) {
-        CTxIn &txin = mergedTx.vin[i];
+    // Collect spent outputs for BIP341 sighash
+    std::vector<CTxOut> spent_outputs;
+    spent_outputs.reserve(txConst.vin.size());
+    for (const CTxIn &txin : txConst.vin) {
         const Coin &coin = view.AccessCoin(txin.prevout);
         if (coin.IsSpent()) {
             throw JSONRPCError(RPC_VERIFY_ERROR,
                                "Input not found or already spent");
         }
+        spent_outputs.push_back(coin.GetTxOut());
+    }
+    const PrecomputedTransactionData txdata(txConst, std::move(spent_outputs));
+    // Sign what we can:
+    for (size_t i = 0; i < mergedTx.vin.size(); i++) {
+        CTxIn &txin = mergedTx.vin[i];
         SignatureData sigdata;
 
-        const CTxOut &txout = coin.GetTxOut();
+        const CTxOut &txout = txdata.m_spent_outputs[i];
 
         // ... and merge in other signatures:
         for (const CMutableTransaction &txv : txVariants) {
@@ -803,10 +810,10 @@ static UniValue combinerawtransaction(const Config &config,
                 sigdata.MergeSignatureData(DataFromTransaction(txv, i, txout));
             }
         }
-        ProduceSignature(
-            DUMMY_SIGNING_PROVIDER,
-            MutableTransactionSignatureCreator(&mergedTx, i, txout.nValue),
-            txout.scriptPubKey, sigdata);
+        ProduceSignature(DUMMY_SIGNING_PROVIDER,
+                         MutableTransactionSignatureCreator(
+                             &mergedTx, i, txout.nValue, SigHashType(), txdata),
+                         txout.scriptPubKey, sigdata);
 
         UpdateInput(txin, sigdata);
     }
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 4a2029e41..cff46e754 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -16,9 +16,10 @@ typedef std::vector<uint8_t> valtype;
 
 MutableTransactionSignatureCreator::MutableTransactionSignatureCreator(
     const CMutableTransaction *txToIn, unsigned int nInIn,
-    const Amount &amountIn, SigHashType sigHashTypeIn)
+    const Amount &amountIn, SigHashType sigHashTypeIn,
+    const PrecomputedTransactionData &txdata)
     : txTo(txToIn), nIn(nInIn), amount(amountIn), sigHashType(sigHashTypeIn),
-      checker(txTo, nIn, amountIn) {}
+      checker(txTo, nIn, amountIn, txdata) {}
 
 bool MutableTransactionSignatureCreator::CreateSig(
     const SigningProvider &provider, std::vector<uint8_t> &vchSig,
@@ -346,15 +347,18 @@ void SignatureData::MergeSignatureData(SignatureData sigdata) {
                       std::make_move_iterator(sigdata.signatures.end()));
 }
 
-bool SignSignature(const SigningProvider &provider, const CScript &fromPubKey,
+bool SignSignature(const SigningProvider &provider,
+                   const PrecomputedTransactionData &txdata,
                    CMutableTransaction &txTo, unsigned int nIn,
-                   const Amount amount, SigHashType sigHashType) {
+                   SigHashType sigHashType) {
     assert(nIn < txTo.vin.size());
 
-    MutableTransactionSignatureCreator creator(&txTo, nIn, amount, sigHashType);
+    MutableTransactionSignatureCreator creator(
+        &txTo, nIn, txdata.m_spent_outputs[nIn].nValue, sigHashType, txdata);
 
     SignatureData sigdata;
-    bool ret = ProduceSignature(provider, creator, fromPubKey, sigdata);
+    bool ret = ProduceSignature(
+        provider, creator, txdata.m_spent_outputs[nIn].scriptPubKey, sigdata);
     UpdateInput(txTo.vin.at(nIn), sigdata);
     return ret;
 }
@@ -365,10 +369,19 @@ bool SignSignature(const SigningProvider &provider, const CTransaction &txFrom,
     assert(nIn < txTo.vin.size());
     CTxIn &txin = txTo.vin[nIn];
     assert(txin.prevout.GetN() < txFrom.vout.size());
-    const CTxOut &txout = txFrom.vout[txin.prevout.GetN()];
 
-    return SignSignature(provider, txout.scriptPubKey, txTo, nIn, txout.nValue,
-                         sigHashType);
+    // Collect spent outputs for BIP341 sighash
+    std::vector<CTxOut> spent_outputs;
+    for (const CTxIn &spending_txin : txTo.vin) {
+        // Make sure the input is spending an output from txFrom
+        assert(spending_txin.prevout.GetTxId() == txFrom.GetId());
+        // Make sure the output of txFrom actually exists
+        assert(spending_txin.prevout.GetN() < txFrom.vout.size());
+        spent_outputs.push_back(txFrom.vout[spending_txin.prevout.GetN()]);
+    }
+
+    const PrecomputedTransactionData txdata(txTo, std::move(spent_outputs));
+    return SignSignature(provider, txdata, txTo, nIn, sigHashType);
 }
 
 namespace {
@@ -444,25 +457,38 @@ bool SignTransaction(CMutableTransaction &mtx, const SigningProvider *keystore,
     // Use CTransaction for the constant parts of the
     // transaction to avoid rehashing.
     const CTransaction txConst(mtx);
-    // Sign what we can:
+    std::vector<CTxOut> spent_outputs(mtx.vin.size());
+    std::vector<bool> has_coin(mtx.vin.size());
     for (size_t i = 0; i < mtx.vin.size(); i++) {
-        CTxIn &txin = mtx.vin[i];
-        auto coin = coins.find(txin.prevout);
+        auto coin = coins.find(mtx.vin[i].prevout);
         if (coin == coins.end() || coin->second.IsSpent()) {
             input_errors[i] = "Input not found or already spent";
+            // Just to be sure, sign for unspendable output
+            spent_outputs[i].scriptPubKey = CScript() << OP_RETURN;
+            continue;
+        }
+        spent_outputs[i] = coin->second.GetTxOut();
+        has_coin[i] = true;
+    }
+    const PrecomputedTransactionData txdata(mtx, std::move(spent_outputs));
+
+    // Sign what we can:
+    for (size_t i = 0; i < mtx.vin.size(); i++) {
+        if (!has_coin[i]) {
             continue;
         }
-        const CScript &prevPubKey = coin->second.GetTxOut().scriptPubKey;
-        const Amount amount = coin->second.GetTxOut().nValue;
+        CTxIn &txin = mtx.vin[i];
+        const CTxOut &coinTxOut = txdata.m_spent_outputs[i];
+        const CScript &prevPubKey = coinTxOut.scriptPubKey;
+        const Amount amount = coinTxOut.nValue;
 
-        SignatureData sigdata =
-            DataFromTransaction(mtx, i, coin->second.GetTxOut());
+        SignatureData sigdata = DataFromTransaction(mtx, i, coinTxOut);
         // Only sign SIGHASH_SINGLE if there's a corresponding output:
         if ((sigHashType.getBaseType() != BaseSigHashType::SINGLE) ||
             (i < mtx.vout.size())) {
             ProduceSignature(*keystore,
-                             MutableTransactionSignatureCreator(&mtx, i, amount,
-                                                                sigHashType),
+                             MutableTransactionSignatureCreator(
+                                 &mtx, i, amount, sigHashType, txdata),
                              prevPubKey, sigdata);
         }
 
diff --git a/src/script/sign.h b/src/script/sign.h
index f2841867e..ed6cc5002 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -45,7 +45,8 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
 public:
     MutableTransactionSignatureCreator(
         const CMutableTransaction *txToIn, unsigned int nInIn,
-        const Amount &amountIn, SigHashType sigHashTypeIn = SigHashType());
+        const Amount &amountIn, SigHashType sigHashTypeIn,
+        const PrecomputedTransactionData &txdata);
     const BaseSignatureChecker &Checker() const override { return checker; }
     bool CreateSig(const SigningProvider &provider,
                    std::vector<uint8_t> &vchSig, const CKeyID &keyid,
@@ -173,9 +174,10 @@ bool ProduceSignature(const SigningProvider &provider,
                       const CScript &scriptPubKey, SignatureData &sigdata);
 
 /** Produce a script signature for a transaction. */
-bool SignSignature(const SigningProvider &provider, const CScript &fromPubKey,
+bool SignSignature(const SigningProvider &provider,
+                   const PrecomputedTransactionData &txdata,
                    CMutableTransaction &txTo, unsigned int nIn,
-                   const Amount amount, SigHashType sigHashType);
+                   SigHashType sigHashType);
 bool SignSignature(const SigningProvider &provider, const CTransaction &txFrom,
                    CMutableTransaction &txTo, unsigned int nIn,
                    SigHashType sigHashType);
diff --git a/src/test/descriptor_tests.cpp b/src/test/descriptor_tests.cpp
index 745f4532d..909b1199a 100644
--- a/src/test/descriptor_tests.cpp
+++ b/src/test/descriptor_tests.cpp
@@ -265,6 +265,16 @@ void DoCheck(const std::string &prv, const std::string &pub, int flags,
                             script_provider_cached1.origins);
             }
 
+            std::vector<CTxOut> spent_outputs;
+            for (const CScript &spk : spks) {
+                spent_outputs.push_back({1 * COIN, spk});
+            }
+            CMutableTransaction spend;
+            spend.vin.resize(1);
+            spend.vout.resize(1);
+            const PrecomputedTransactionData txdata(spend,
+                                                    std::move(spent_outputs));
+
             // For each of the produced scripts, verify solvability, and when
             // possible, try to sign a transaction spending it.
             for (size_t n = 0; n < spks.size(); ++n) {
@@ -274,13 +284,9 @@ void DoCheck(const std::string &prv, const std::string &pub, int flags,
                     (flags & UNSOLVABLE) == 0);
 
                 if (flags & SIGNABLE) {
-                    CMutableTransaction spend;
-                    spend.vin.resize(1);
-                    spend.vout.resize(1);
                     BOOST_CHECK_MESSAGE(
-                        SignSignature(Merge(keys_priv, script_provider),
-                                      spks[n], spend, 0, 1 * COIN,
-                                      SigHashType().withForkId()),
+                        SignSignature(Merge(keys_priv, script_provider), txdata,
+                                      spend, 0, SigHashType().withForkId()),
                         prv);
                 }
 
diff --git a/src/test/fuzz/script_sign.cpp b/src/test/fuzz/script_sign.cpp
index d676c0494..2b2e8288e 100644
--- a/src/test/fuzz/script_sign.cpp
+++ b/src/test/fuzz/script_sign.cpp
@@ -102,7 +102,7 @@ void test_one_input(const std::vector<uint8_t> &buffer) {
         if (mutable_transaction && tx_out &&
             mutable_transaction->vin.size() > n_in) {
             SignatureData signature_data_1 =
-                DataFromTransaction(*mutable_transaction, n_in, *tx_out);
+                DataFromTransaction(*mutable_transaction, n_in, {*tx_out});
             CTxIn input;
             UpdateInput(input, signature_data_1);
             const CScript script = ConsumeScript(fuzzed_data_provider);
@@ -136,7 +136,7 @@ void test_one_input(const std::vector<uint8_t> &buffer) {
                 MutableTransactionSignatureCreator signature_creator{
                     &tx_to, n_in, ConsumeMoney(fuzzed_data_provider),
                     SigHashType(
-                        fuzzed_data_provider.ConsumeIntegral<uint32_t>())};
+                        fuzzed_data_provider.ConsumeIntegral<uint32_t>()), {}};
                 std::vector<uint8_t> vch_sig;
                 CKeyID address;
                 if (fuzzed_data_provider.ConsumeBool()) {
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index f15f5bd29..efda78c94 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -2351,11 +2351,14 @@ SignatureData CombineSignatures(const CTxOut &txout,
                                 const CMutableTransaction &tx,
                                 const SignatureData &scriptSig1,
                                 const SignatureData &scriptSig2) {
+    assert(tx.vin.size() == 1);
     SignatureData data;
     data.MergeSignatureData(scriptSig1);
     data.MergeSignatureData(scriptSig2);
+    const PrecomputedTransactionData txdata(tx, {txout});
     ProduceSignature(DUMMY_SIGNING_PROVIDER,
-                     MutableTransactionSignatureCreator(&tx, 0, txout.nValue),
+                     MutableTransactionSignatureCreator(&tx, 0, txout.nValue,
+                                                        SigHashType(), txdata),
                      txout.scriptPubKey, data);
     return data;
 }
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index f6a3507d1..73dad9c68 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -38,6 +38,7 @@
 
 #include <map>
 #include <string>
+#include <algorithm>
 
 typedef std::vector<uint8_t> valtype;
 
@@ -424,12 +425,17 @@ BOOST_AUTO_TEST_CASE(test_big_transaction) {
         mtx.vout.emplace_back(1000 * SATOSHI, CScript() << OP_1);
     }
 
-    // sign all inputs
-    for (size_t i = 0; i < mtx.vin.size(); i++) {
-        bool hashSigned =
-            SignSignature(keystore, scriptPubKey, mtx, i, 1000 * SATOSHI,
-                          sigHashes.at(i % sigHashes.size()));
-        BOOST_CHECK_MESSAGE(hashSigned, "Failed to sign test transaction");
+    {
+        const CTxOut output{1000 * SATOSHI, scriptPubKey};
+        std::vector<CTxOut> spent_outputs(mtx.vin.size());
+        std::fill_n(spent_outputs.begin(), spent_outputs.size(), output);
+        const PrecomputedTransactionData txdata(mtx, std::move(spent_outputs));
+        // sign all inputs
+        for (size_t i = 0; i < mtx.vin.size(); i++) {
+            bool hashSigned = SignSignature(keystore, txdata, mtx, i,
+                                            sigHashes.at(i % sigHashes.size()));
+            BOOST_CHECK_MESSAGE(hashSigned, "Failed to sign test transaction");
+        }
     }
 
     CTransaction tx(mtx);
@@ -478,10 +484,11 @@ SignatureData CombineSignatures(const CMutableTransaction &input1,
     SignatureData sigdata;
     sigdata = DataFromTransaction(input1, 0, tx->vout[0]);
     sigdata.MergeSignatureData(DataFromTransaction(input2, 0, tx->vout[0]));
-    ProduceSignature(
-        DUMMY_SIGNING_PROVIDER,
-        MutableTransactionSignatureCreator(&input1, 0, tx->vout[0].nValue),
-        tx->vout[0].scriptPubKey, sigdata);
+    const PrecomputedTransactionData txdata(*tx, {tx->vout[0]});
+    ProduceSignature(DUMMY_SIGNING_PROVIDER,
+                     MutableTransactionSignatureCreator(
+                         &input1, 0, tx->vout[0].nValue, SigHashType(), txdata),
+                     tx->vout[0].scriptPubKey, sigdata);
     return sigdata;
 }
 
diff --git a/src/test/txvalidationcache_tests.cpp b/src/test/txvalidationcache_tests.cpp
index c5485b4ee..5eb90ce40 100644
--- a/src/test/txvalidationcache_tests.cpp
+++ b/src/test/txvalidationcache_tests.cpp
@@ -431,23 +431,25 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, TestChain100Setup) {
         tx.vout[0].nValue = 22 * CENT;
         tx.vout[0].scriptPubKey = p2pk_scriptPubKey;
 
+        const PrecomputedTransactionData precomputed_txdata(
+            tx, {spend_tx.vout[0], spend_tx.vout[3]});
         // Sign
         {
             SignatureData sigdata;
+            MutableTransactionSignatureCreator creator(
+                &tx, 0, 11 * CENT, SigHashType().withForkId(),
+                precomputed_txdata);
             BOOST_CHECK(ProduceSignature(
-                keystore,
-                MutableTransactionSignatureCreator(&tx, 0, 11 * CENT,
-                                                   SigHashType().withForkId()),
-                spend_tx.vout[0].scriptPubKey, sigdata));
+                keystore, creator, spend_tx.vout[0].scriptPubKey, sigdata));
             UpdateInput(tx.vin[0], sigdata);
         }
         {
             SignatureData sigdata;
+            MutableTransactionSignatureCreator creator(
+                &tx, 1, 11 * CENT, SigHashType().withForkId(),
+                precomputed_txdata);
             BOOST_CHECK(ProduceSignature(
-                keystore,
-                MutableTransactionSignatureCreator(&tx, 1, 11 * CENT,
-                                                   SigHashType().withForkId()),
-                spend_tx.vout[3].scriptPubKey, sigdata));
+                keystore, creator, spend_tx.vout[3].scriptPubKey, sigdata));
             UpdateInput(tx.vin[1], sigdata);
         }
 
-- 
2.39.5 (Apple Git-154)

